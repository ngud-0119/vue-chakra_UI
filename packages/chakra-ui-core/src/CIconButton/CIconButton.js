/**
 * Hey! Welcome to @chakra-ui/vue IconButton
 *
 * CIconButton is used to render icons that support
 * click interactions.
 *
 * CIconButton composes the CButton component, except that it renders only an icon.
 *
 * @see Docs     https://vue.chakra-ui.com/iconbutton
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue/blob/master/packages/chakra-ui-core/src/CIconButton/CIconButton.js
 */

import styleProps from '../config/props'
import { forwardProps } from '../utils'
import { buttonProps } from '../CButton/utils/button.props'

import CButton from '../CButton'
import CIcon from '../CIcon'
import CBox from '../CBox'

const baseStyles = {
  display: 'inline-flex',
  appearance: 'none',
  alignItems: 'center',
  justifyContent: 'center',
  transition: 'all 250ms',
  userSelect: 'none',
  position: 'relative',
  whiteSpace: 'nowrap',
  verticalAlign: 'middle',
  lineHeight: '1.2',
  outline: 'none'
}

/**
 * CIconButton component
 *
 * CIconButton is used to render icons that support
 * click interactions
 *
 * @extends CButton
 * @see Docs https://vue.chakra-ui.com/iconbutton
 */
const CIconButton = {
  name: 'CIconButton',
  inject: ['$chakraTheme', '$chakraColorMode'],
  props: {
    icon: {
      type: [String]
    },
    isRound: {
      type: [Boolean]
    },
    ariaLabel: {
      type: [String],
      required: true
    },
    ...buttonProps,
    ...styleProps
  },
  render (h) {
    const { isFullWidth, leftIcon, rightIcon, loadingText, ...props } = this.$props

    return h(CButton, {
      props: {
        p: 0,
        rounded: this.isRound ? 'full' : 'md',
        size: this.size,
        ...forwardProps(props)
      },
      attrs: {
        'aria-label': this.ariaLabel,
        'data-chakra-component': 'CIconButton'
      },
      on: {
        click: e => this.$emit('click', e)
      }
    },
    [typeof this.icon === 'string'
      ? h(CIcon, {
        props: {
          ...baseStyles,
          name: this.icon,
          color: 'currentColor',
          mb: '2px',
          size: '1em'
        },
        attrs: {
          focusable: false,
          'aria-hidden': true
        }
      })
      : h(CBox, {
        props: {
          as: this.icon,
          color: 'currentColor'
        },
        attrs: {
          focusable: true
        }
      })]
    )
  }
}

export default CIconButton
