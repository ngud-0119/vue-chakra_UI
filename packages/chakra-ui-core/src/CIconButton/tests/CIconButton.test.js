import CIconButton from '..'
import { render, userEvent } from '@/tests/test-utils'

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { CIconButton },
    template: `<CIconButton aria-label="Phone" variant-color="blue" icon="phone" data-testid="btn" ${inlineAttrs} />`,
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const { asFragment } = renderComponent()

  expect(asFragment()).toMatchSnapshot()
})

it('should display spinner and hide the icon', () => {
  const inlineAttrs = 'isLoading'
  const { container, getByTestId } = renderComponent({ inlineAttrs })

  const button = getByTestId('btn')

  expect(button).toHaveAttribute('disabled')
  expect(button).toHaveAttribute('aria-disabled', 'true')

  expect(container.querySelector('[data-chakra-component=CSpinner]')).toBeInTheDocument()
  expect(button).toHaveStyle('opacity: 0.4')
  expect(container.querySelector('button > div')).toBeInTheDocument()
})

it('should change icon', () => {
  const { asFragment } = renderComponent({
    template: `
    <CIconButton aria-label="Phone" variant-color="blue" icon="add" data-testid="btn" />`
  })

  expect(asFragment()).toMatchSnapshot()
})

it('should emit "click" event', () => {
  const handleClick = jest.fn()
  const { getByTestId } = renderComponent({
    template: '<CIconButton aria-label="Phone" @click="handleClick" variant-color="blue" icon="add" data-testid="btn" />',
    methods: {
      handleClick
    }
  })

  const button = getByTestId('btn')

  userEvent.click(button)
  expect(button).toHaveFocus()
  expect(handleClick).toHaveBeenCalled()
})
