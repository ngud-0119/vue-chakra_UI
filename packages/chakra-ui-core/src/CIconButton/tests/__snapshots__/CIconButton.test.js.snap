// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change icon 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: 2px solid transparent;
  outline-offset: 2px;
  font-weight: var(--chakra-fontWeights-bold);
  border-radius: var(--chakra-radii-md);
  height: var(--chakra-sizes-10);
  min-width: var(--chakra-sizes-10);
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  background: var(--chakra-colors-blue-400);
  color: var(--chakra-colors-white);
  padding: var(--chakra-space-0);
}

.emotion-0:focus,
.emotion-0[data-focus] {
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: var(--chakra-colors-blue-500);
}

.emotion-0:active,
.emotion-0[data-active] {
  background: var(--chakra-colors-blue-600);
}

.emotion-0[disabled],
.emotion-0[aria-disabled=true],
.emotion-0[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-1 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.emotion-1:not(:root) {
  overflow: hidden;
}

.emotion-2 {
  width: 1em;
  height: 1em;
  display: inline-block;
  vertical-align: middle;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: 2px solid transparent;
  outline-offset: 2px;
  color: currentColor;
  margin-bottom: 2px;
}

<button
    aria-label="Phone"
    class="emotion-0"
    data-chakra-component="CIconButton"
    data-testid="btn"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="emotion-1 emotion-2 emotion-3"
      data-chakra-component="CIcon"
      role="presentation"
      viewBox="0 0 24 24"
    >
      
    
      <path
        d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
        fill="currentColor"
      />
      
    
    </svg>
  </button>
</DocumentFragment>
`;

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: 2px solid transparent;
  outline-offset: 2px;
  font-weight: var(--chakra-fontWeights-bold);
  border-radius: var(--chakra-radii-md);
  height: var(--chakra-sizes-10);
  min-width: var(--chakra-sizes-10);
  font-size: var(--chakra-fontSizes-md);
  -webkit-padding-start: var(--chakra-space-4);
  padding-inline-start: var(--chakra-space-4);
  -webkit-padding-end: var(--chakra-space-4);
  padding-inline-end: var(--chakra-space-4);
  background: var(--chakra-colors-blue-400);
  color: var(--chakra-colors-white);
  padding: var(--chakra-space-0);
}

.emotion-0:focus,
.emotion-0[data-focus] {
  outline: 2px solid transparent;
  outline-offset: 2px;
  box-shadow: var(--chakra-shadows-outline);
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: var(--chakra-colors-blue-500);
}

.emotion-0:active,
.emotion-0[data-active] {
  background: var(--chakra-colors-blue-600);
}

.emotion-0[disabled],
.emotion-0[aria-disabled=true],
.emotion-0[data-disabled] {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: var(--chakra-shadows-none);
}

.emotion-1 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
}

.emotion-1:not(:root) {
  overflow: hidden;
}

.emotion-2 {
  width: 1em;
  height: 1em;
  display: inline-block;
  vertical-align: middle;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: 2px solid transparent;
  outline-offset: 2px;
  color: currentColor;
  margin-bottom: 2px;
}

<button
    aria-label="Phone"
    class="emotion-0"
    data-chakra-component="CIconButton"
    data-testid="btn"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="emotion-1 emotion-2 emotion-3"
      data-chakra-component="CIcon"
      role="presentation"
      viewBox="0 0 14 14"
    >
      
    
      <path
        d="M2.20731,0.0127209 C2.1105,-0.0066419 1.99432,-0.00664663 1.91687,0.032079 C0.871279,0.438698 0.212942,1.92964 0.0580392,2.95587 C-0.426031,6.28627 2.20731,9.17133 4.62766,11.0689 C6.77694,12.7534 10.9012,15.5223 13.3409,12.8503 C13.6507,12.5211 14.0186,12.037 13.9993,11.553 C13.9412,10.7397 13.186,10.1588 12.6051,9.71349 C12.1598,9.38432 11.2304,8.47427 10.6495,8.49363 C10.1267,8.51299 9.79754,9.05515 9.46837,9.38432 L8.88748,9.96521 C8.79067,10.062 7.55145,9.24878 7.41591,9.15197 C6.91248,8.8228 6.4284,8.45491 6.00242,8.04829 C5.57644,7.64167 5.18919,7.19632 4.86002,6.73161 C4.7632,6.59607 3.96933,5.41495 4.04678,5.31813 C4.04678,5.31813 4.72448,4.58234 4.91811,4.2919 C5.32473,3.67229 5.63453,3.18822 5.16982,2.45243 C4.99556,2.18135 4.78257,1.96836 4.55021,1.73601 C4.14359,1.34875 3.73698,0.942131 3.27227,0.612963 C3.02055,0.419335 2.59457,0.0708094 2.20731,0.0127209 Z"
        fill="currentColor"
      />
      
    
    </svg>
  </button>
</DocumentFragment>
`;
